### ISSUE TITLE
Wrong message R1721 unnecessary-comprehension

### ISSUE THREAD

**Issue by socketpair** [PyCQA/pylint#3452]
```python
x = {k: v for (k, v) in (line.split('=', maxsplit=1) for line in stdout.splitlines())}
```
generates this. So, in order to refactor this I have to use typing.cast() for mypy not to bark:

```python
x = dict(cast(Tuple[str, str], line.split('=', maxsplit=1)) for line in stdout.splitlines()))
```
which looks uglier, and also I can't use dict comprehension, as I wanted to.

**Comment by PCManticore** [comment]
So this check recommends you to use the `dict()` constructor instead. I'm not entirely sure what's `mypy` got to do with it? Surely you can also use `x: Dict[str, str] = dict(....)` instead? I don't think this is a valid issue at all.

**Comment by socketpair** [comment]
No, the problem in another thing. `cast` solves another problem: `dict()` constructor wants `Tuple[zzz, yyy]`, But I pass `List[str]`.
